File: Rust/23.merge-k-sorted-lists.rs
/*
 * @lc app=leetcode id=23 lang=rust
 *
 * [23] Merge k Sorted Lists
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
use std::cmp::Ordering;
use std::collections::BinaryHeap; // BinaryHeap is Max Heap in default

impl PartialOrd<ListNode> for ListNode {
    fn partial_cmp(&self, other: &ListNode) -> Option<Ordering> {
        other.val.partial_cmp(&self.val)
    }
}

impl Ord for ListNode {
    fn cmp(&self, other: &Self) -> Ordering {
        other.val.cmp(&self.val)
    }
}

impl Solution {
    pub fn merge_k_lists(lists: Vec<Option<Box<ListNode>>>) -> Option<Box<ListNode>> {
        let mut res = Box::new(ListNode::new(0));
        if lists.len() == 0 {
            return None;
        }
        let mut minHeap = BinaryHeap::new();
        for mut list in lists {
            if list.is_some() {
                minHeap.push(list.take()?); // Storing as Option<Box<ListNode>>
            }
        }
        let mut cur = &mut res;
        while !minHeap.is_empty() {
            cur.next = minHeap.pop(); // Return as Option<Box<ListNode>>
            cur = cur.next.as_mut()?;
            if cur.next.is_some() {
                minHeap.push(cur.next.take()?);
            }
        }
        res.next
    }
}
// @lc code=end

File: JS/23.merge-k-sorted-lists.js
/*
 * @lc app=leetcode id=23 lang=javascript
 *
 * [23] Merge k Sorted Lists
 */

// @lc code=start

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
/*
可以用merge sort的算法divide & conquer，分到最小两个数组再合并起来，follow up 可以用
heap存lists.length个头结点
*/
var mergeKLists = function(lists) {
  if(lists === null || lists.length === 0) return null;
  return merge(lists, 0, lists.length - 1);
  
  function merge(lists, left, right) {
    if(left === right) {
      return lists[left];
    } else if(left < right) {
      let mid = (left + right) >>> 1;
      let l = merge(lists, left, mid);
      let r = merge(lists, mid + 1, right);
      return mergeTwoSortedLists(l, r);
    } else {
      return null;
    }
  }
  
  function mergeTwoSortedLists(l1, l2) {
    let dummy = { next: null };
    let cur = dummy;
    while(l1 !== null && l2 !== null) {
      if(l1.val < l2.val) {
        cur.next = l1;
        l1 = l1.next;
      } else {
        cur.next = l2;
        l2 = l2.next;
      }
      cur = cur.next;
    }
    if(l1 !== null) cur.next = l1;
    if(l2 !== null) cur.next = l2;
    return dummy.next;
  }
};
// @lc code=end

